src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceCacheKey.cs

```diff
@@ -5,7 +5,7 @@
 
 namespace Microsoft.Extensions.DependencyInjection.ServiceLookup
 {
-    internal struct ServiceCacheKey: IEquatable<ServiceCacheKey>
+    internal readonly struct ServiceCacheKey : IEquatable<ServiceCacheKey>
```

Comment:
Not directly related to your changes, but the HashCode calculation will throw for the static ``Empty``, the Type is null then, and GetHashCode will throw a null ref exception.
I suggest changing it to ``HashCode.Combine(Type, Slot)`` which takes care of the null

Reply:
Will do that in a different PR.